File: epi_judge_python/bst_from_preorder.py
from typing import List, Optional

from bst_node import BstNode
from test_framework import generic_test


def rebuild_bst_from_preorder(preorder_sequence: List[int]
                              ) -> Optional[BstNode]:

    def rebuild_bst(seq):

        if len(seq) == 0:
            return None

        """
        so each subtree 
        """
        rootval = seq[0]
        idx = 1
        while idx < len(seq) and seq[0] > seq[idx]:
            idx += 1

        nodeL = rebuild_bst(seq[1:idx])
        nodeR = rebuild_bst(seq[idx:])

        return BstNode(rootval, nodeL, nodeR)

    return rebuild_bst(preorder_sequence)


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('bst_from_preorder.py',
                                       'bst_from_preorder.tsv',
                                       rebuild_bst_from_preorder))

File: epi_judge_cpp/bst_from_preorder.cc
#include <memory>

#include "bst_node.h"
#include "test_framework/generic_test.h"
using std::unique_ptr;
using std::vector;
// using std::next;
unique_ptr<BstNode<int>> RebuildBSTFromPreorder(
    const vector<int> &preorder_sequence)
{
  if (preorder_sequence.size() == 0)
  {
    return nullptr;
  }
  // TODO - you fill in here.
  auto transition = preorder_sequence.cbegin();
  auto currVal = std::next(transition);
  while (currVal != preorder_sequence.end() && *transition > *currVal)
  {

    currVal = std::next(currVal);
  }
  // std::next(transition);

  const vector<int> leftSubTree(transition + 1, currVal);
  const vector<int> rightSubTree(currVal, preorder_sequence.end());

  auto lnode = RebuildBSTFromPreorder(leftSubTree);
  auto rnode = RebuildBSTFromPreorder(rightSubTree);
  auto newBstNode = std::make_unique<BstNode<int>>(*transition, lnode, rnode);
  return newBstNode;
}

int main(int argc, char *argv[])
{
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"preorder_sequence"};
  return GenericTestMain(args, "bst_from_preorder.cc", "bst_from_preorder.tsv",
                         &RebuildBSTFromPreorder, DefaultComparator{},
                         param_names);
}
